### DISCUSSION TITLE
Invalid name of some module in production mode

### DISCUSSION THREAD

**Discussion by bupy7** [webpack/webpack#16513] [Q&A]
# Bug report

**What is the current behavior?**

Sometimes without any reasons my project do not works properly in production mode. If I build up my project in development mode - everything is perfect, no any errors. But in production mode I have some weird problem.

Sometimes after build up I see an error like this in the Developer Console:

<img width="577" alt="Screenshot at Nov 24 18-43-45" src="https://user-images.githubusercontent.com/5145037/203800599-ea3f62e0-47b3-4de5-85c4-30f660cbe29c.png">

As you see, the d8498 is not defined. I have looked into the file 286_6660b.js and saw this:

```
/* harmony import */ var _code_editor__WEBPACK_IMPORTED_MODULE_5__ = __webpack_require__(d8498);
```

The name of module just not in enclosed quotes, because `d8498` is not valid name for this situation. And if I fix this name adding enclosed quotes the error is gone. And it's not the most weird. I'v found the component of the module `code_editor` and just renamed the directory `code-editor` adding `_` in the beginning of name to make it `_code-editor`. After it the error is gone too. Very weird. 

**If the current behavior is a bug, please provide the steps to reproduce.**

I don't know how to reproduce it, because it has periodic character. Sometimes I fix with renaming files/directories, sometimes deleting JSDoc tags. It always the same error regardless to whether use minimization in Webpack.

**What is the expected behavior?**

I expect a valid number name for all modules after building in production mode.

**Other relevant information:**
webpack version:  5.75
Node.js version: 16.13.2
Operating System:  Linux Debian 10 Buster

**Comment by alexander-akait** [comment]
As you can see this error from the browser extension, sounds like you have race condition somewhere, we need as minimum reproducible test repo, otherwise we really can't help you, can you provode your configuration?

**Reply by bupy7** [reply]
Do you mean Webpack config file?

**Reply by alexander-akait** [reply]
Yes

**Reply by bupy7** [reply]
```js
// noinspection NodeCoreCodingAssistance

const path = require('path');
const webpack = require('webpack');
const DefinePlugin = webpack.DefinePlugin;
const IgnorePlugin = webpack.IgnorePlugin;
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const TerserPlugin = require('terser-webpack-plugin');
const CssMinimizerPlugin = require('css-minimizer-webpack-plugin');

module.exports = (env) => {
    // noinspection JSUnresolvedVariable
    const APP_CONFIG = {
        isProd: env.NODE_ENV === 'production',
        isDev: env.NODE_ENV === 'development',
        appUrl: env.APP_URL,
        oldAppUrl: env.OLD_APP_URL
    };

    const ASSETS_DIR = 'webpack-assets';

    return {
        mode: env.NODE_ENV,
        devtool: !APP_CONFIG.isProd ? 'eval-cheap-module-source-map' : false,
        context: __dirname,
        entry: {
            admin: './src/admin/index.js',
            print: './src/print/index.js'
        },
        output: {
            filename: `${ASSETS_DIR}/[name]_[contenthash:5].js`,
            path: path.resolve(__dirname, 'public'),
            publicPath: '/',
            clean: true
        },
        module: {
            rules: [
                {
                    test: /\.css/,
                    use: [
                        MiniCssExtractPlugin.loader,
                        {
                            loader: 'css-loader',
                            options: {
                                sourceMap: false
                            }
                        }
                    ]
                },
                {
                    test: /\.less$/,
                    use: [
                        MiniCssExtractPlugin.loader,
                        {
                            loader: 'css-loader',
                            options: {
                                sourceMap: false
                            }
                        },
                        {
                            loader: 'less-loader',
                            options: {
                                lessOptions: {
                                    math: 'strict' // https://lesscss.org/usage/#less-options-math
                                }
                            }
                        }
                    ],
                },
                {
                    test: /\.(sass|scss)$/,
                    use: [
                        MiniCssExtractPlugin.loader,
                        {
                            loader: 'css-loader',
                            options: {
                                sourceMap: false
                            }
                        },
                        'sass-loader'
                    ],
                },
                {
                    test: /\.(eot|woff|woff2|ttf)$/,
                    type: 'asset/resource',
                    generator: {
                        filename: `${ASSETS_DIR}/[name]_[contenthash:5][ext]`
                    }
                },
                {
                    test: /\.(svg|png|jpg|jpeg|gif)$/,
                    type: 'asset/resource',
                    generator: {
                        filename: `${ASSETS_DIR}/[name]_[contenthash:5][ext]`
                    }
                },
                {
                    test: /\.js$/,
                    exclude: /node_modules/,
                    use: 'babel-loader'
                }
            ]
        },
        plugins: [
            function () {
                this.hooks.compile.tap('watch-run', () => {
                    console.log("\n\x1b[36m" + 'Begin compile at ' + new Date() + '\x1b[0m ');
                });
            },
            new HtmlWebpackPlugin({
                inject: false,
                filename: 'index.html',
                template: 'template/index.hbs',
                chunks: ['admin'],
                // below are custom options
                appName: 'admin'
            }),
            new HtmlWebpackPlugin({
                inject: false,
                filename: 'print.html',
                template: 'template/index.hbs',
                chunks: ['print'],
                // below are custom options
                appName: 'print',
            }),
            new MiniCssExtractPlugin({
                filename: `${ASSETS_DIR}/[name]_[contenthash:5].css`
            }),
            new DefinePlugin({
                __MYAPP: JSON.stringify(APP_CONFIG)
            }),
            new IgnorePlugin({resourceRegExp: /^\.\/locale$/, contextRegExp: /moment$/})
        ],
        watchOptions: {
            poll: true
        },
        optimization: {
            concatenateModules: false, // https://github.com/webpack/webpack/issues/959
            minimizer: [
                new TerserPlugin({
                    parallel: true
                }),
                new CssMinimizerPlugin({
                    parallel: true
                })
            ],
            splitChunks: {
                chunks: 'all'
            }
        },
        resolve: {
            alias: {
                'admin': path.resolve(__dirname, 'src/admin/'),
                'common': path.resolve(__dirname, 'src/common/'),
                'order': path.resolve(__dirname, 'src/order/'),
                'print': path.resolve(__dirname, 'src/print/'),
                'storage': path.resolve(__dirname, 'src/storage/')
            }
        }
    };
};
```

**Reply by bupy7** [reply]
Can race condition occurs if I use both alias name and relative path at import path inside the same directory? For example:

```js
// src/admin/example.js

import SomeFile from 'admin/some-file';
```

**Reply by alexander-akait** [reply]
> Can race condition occurs if I use both alias name and relative path at import path inside the same directory?

No

Can you check versions of your extensions for your browser (I think you need latest)?

**Reply by bupy7** [reply]
I have some browser extensions, but how is this related? The error occurs even if there are no extensions. May be I don't understand question... In the project `package.json` are no any browser extensions at all.

**Reply by bupy7** [reply]
I see! Error had caught with browser extension React Developer Tool. But below was the same unhandled error. And now I don't know how to reproduce the error to make new screenshot.

**Reply by alexander-akait** [reply]
Yeah, that is why I asked to update extensions for your browser

**Reply by bupy7** [reply]
I'v made rollback to the day when error is occurred and made new screenshot:

<img width="648" alt="image" src="https://user-images.githubusercontent.com/5145037/203991355-ba5cafcc-7e69-47d7-85ee-b86497347261.png">

(Information for me only. Commit hash: 7c385833)

**Reply by alexander-akait** [reply]
Intresting can you try to set https://webpack.js.org/configuration/output/#outputuniquename to any random value and check it

**Reply by bupy7** [reply]
Nothing changes. The same error.

**Reply by bupy7** [reply]
If I'm adding enclosed quotes for d8498 - everything is working:

```js
/* harmony import */ var _code_editor__WEBPACK_IMPORTED_MODULE_5__ = __webpack_require__('d8498');
```

And without they - the error is occurring.

**Reply by bupy7** [reply]
I found this https://stackoverflow.com/questions/66820405/incorrect-id-generation-for-modules-in-webpack , but suggestion about `terser-webpack-plugin` have not solved my problem.

**Reply by alexander-akait** [reply]
Can you provide your confiuration?

**Reply by bupy7** [reply]
Configuration of what exactly?

**Reply by alexander-akait** [reply]
I mean `webpack.config.js`?

**Reply by bupy7** [reply]
I'v sent you my configuration of Webpack [above](https://github.com/webpack/webpack/discussions/16513#discussioncomment-4236882).

**Reply by alexander-akait** [reply]
Can you change `[contenthash:5]` to `[contenthash:8]`? `5` is very small, you can have overlapping

**Reply by bupy7** [reply]
Yeap! It's looking like fixed my problem. Thanks a lot!

**Reply by alexander-akait** [reply]
When you use very small value (for example `3` you can overlap your hashes with numbers (for example - `100`, is it number in code or hash?), I recommend to use `8` as minimum to avoid collisions)

**Reply by bupy7** [reply]
I see. Thanks! Although "8" do not give 100% guarantees either. But now I know where problem is.

**Comment by phoenixinwater** [comment]
Hi @alexander-akait I have also experienced this issue but the offending hash for me is the same as the `MiniCssExtractPlugin` output contenthash. Would you be able to point to me where in the src, if it is possible, where a conflict between module IDs and loader output hashes could occur? The module IDs are all numeric on init but at some point in the build, that I am struggling to find, one is replaced with a non-numeric hash. Example of 'bad' hashes are `09e92` and `9b802` for css file contenthash length of 4.

**Reply by alexander-akait** [reply]
@phoenixinwater Sorry for delay, is it still valid?
